primitives/trie/src/cache/shared_cache.rs

```diff
@@ -517,23 +616,73 @@ impl<H: Hasher> Clone for SharedTrieCache<H> {
 impl<H: Hasher> SharedTrieCache<H> {
 	/// Create a new [`SharedTrieCache`].
 	pub fn new(cache_size: CacheSize) -> Self {
-		let (node_cache_size, value_cache_size) = match cache_size {
-			CacheSize::Maximum(max) => {
-				// Allocate 20% for the value cache.
-				let value_cache_size_in_bytes = (max as f32 * 0.20) as usize;
-
-				(
-					CacheSize::Maximum(max - value_cache_size_in_bytes),
-					CacheSize::Maximum(value_cache_size_in_bytes),
-				)
-			},
-			CacheSize::Unlimited => (CacheSize::Unlimited, CacheSize::Unlimited),
-		};
+		let total_budget = cache_size.0;
+
+		// Split our memory budget between the two types of caches.
+		let value_cache_budget = (total_budget as f32 * 0.20) as usize; // 20% for the value cache
+		let node_cache_budget = total_budget - value_cache_budget; // 80% for the node cache
+
+		// Split our memory budget between what we'll be holding inline in the map,
+		// and what we'll be holding on the heap.
+		let mut value_cache_inline_budget = (value_cache_budget as f32 * 0.70) as usize;
+		let mut node_cache_inline_budget = (node_cache_budget as f32 * 0.70) as usize;
+
+		// Calculate how much memory the maps will be allowed to hold inline given our budget.
+		let mut value_cache_max_inline_size;
+		loop {
+			// The map is resized in powers of two, so we'll almost certainly blow our inline
+			// budget.
```

Comment:
Why isn't the internal logic changed to tell the map the upper limit and then for the last allocate we allocate up to the limit and not try to allocate by power of two?

Reply:
Ah, good catch! This is a vestigial remain of a previous version of the code; the `memory_usage_for_memory_budget` is guaranteed to return a value that's less or equal to the given budget , so this loop's essentially useless (it will just always run once). Harmless, but useless. I'll remove it.